### PR TITLE
Prepare weights for ilp unwrapping

### LINKED ISSUE #6: Bug in weights for ILP
Hey, I've noticed what seems to be a typo in [line 104 in `core.py`](https://github.com/yoyololicon/kamui/blob/dev/kamui/core.py#L104).
The `c` which you use as argument to `np.tile` is not defined earlier.
I'd appreciate if you could let me know what the correct argument here is so that I can use weights in my computations.

### PR DESCRIPTION
**Pull request by the-mysh**
Hi @yoyololicon, I think I have figured out a way of passing the weights to the ILP unwrapping. 

Please review it; in particular, I was wondering if the rescaling approach and calculating the mean to go from weight-per-phase to weight-per edge is correct.

If it's all right, I think it should resolve #6 which I opened earlier.

### CODE REVIEW

**Comment by the-mysh** [comment]
I've now added a choice of methods to compute an edge weight from a pair of phase weights: choose from mean, max, and min.

File: kamui/utils.py [hunk 1]
```diff
@@ -194,3 +195,49 @@ def get_3d_edges_and_simplices(
...
+    """Prepare weights for `calculate_m` and `calculate_k` functions.
+
+    Assume the weights are the same shape as the phases to be unwrapped.
+
+    Scale the weights from 0 to 1. Pick the weights corresponding to the phase pairs connected by the edges.
+    Compute the mean/max/min (depending on the `merging_method`) of each of those pairs to give a weight for each edge.
+
+    Args:
+        weights         :   Array of weights of shapr corresponding to the original phases array shape.
+        edges           :   Edges connecting the phases. Shape: (N, 2), where N is the number of edges.
```
Comment (yoyololicon) [inline kamui/utils.py:211]: For consistency.
```suggestion
        edges           :   Edges connecting the phases. Shape: (M, 2), where M is the number of edges.
```

File: kamui/utils.py [hunk 2]
```diff
@@ -194,3 +195,49 @@ def get_3d_edges_and_simplices(
...
+        smoothing       :   A positive value between 0 (inclusive) and 1 (not inclusive). This is the minimal value
+                            of the rescaled weights where they are defined. If smoothing > 0, the value of 0 is reserved
+                            for places where the weights are originally NaN. If smoothing == 0, 0 will be used for both
+                            NaN weights and smallest non-NaN ones.
+        merging_method  :   Way of combining two phase weights into a single edge weight.
+
+    Returns:
+        Array of weights for the edges, shape: (N,). Rescaled to (0, 1).
```
Comment (yoyololicon) [inline kamui/utils.py:219]:
```suggestion
        Array of weights for the edges, shape: (M,). Rescaled to (0, 1).
```

**Review by yoyololicon — CHANGES_REQUESTED** [review]
`prepare_weights` looks nice, but it should be the `unwrap_dimensional` to call it instead of directly calling `calculate_k/m`. The weights to the latter are edge weights and should have a shape of `(M,)`, not `(N,)`. 

I suggest the following:
```python
def unwrap_dimensional(
    x: np.ndarray,
    start_pixel: Optional[Union[Tuple[int, int], Tuple[int, int, int]]] = None,
    use_edgelist: bool = False,
    cyclical_axis: Union[int, Tuple[int, int]] = (),
    merging_method: str = 'mean',
    weights = None,
    **kwargs,
) -> np.ndarray:
   """
    Unwrap the phase of a 2-D or 3-D array.

    Parameters
    ----------
    x : 2-D or 3-D np.ndarray
        The phase to be unwrapped.
    start_pixel : (2,) or (3,) tuple
        the reference pixel to start unwrapping.
        Default to (0, 0) for 2-D data and (0, 0, 0) for 3-D data.
    use_edgelist : bool
        Whether to use the edgelist method.
        Default to False.
    cyclical_axis : int or (int, int)
        The axis that is cyclical.
        Default to ().
    weights: 2-D or 3-D np.ndarray
        array of weights for each pixel.
    kwargs : dict
        Other arguments besides `weights` passed to `kamui.unwrap_arbitrary`.

    Returns
    -------
    np.ndarray
        The unwrapped phase of the same shape as x.
    """
    if start_pixel is None:
        start_pixel = (0,) * x.ndim
    assert x.ndim == len(start_pixel), "start_pixel must have the same dimension as x"

    start_i = 0
    for i, s in enumerate(start_pixel):
        start_i *= x.shape[i]
        start_i += s

    if x.ndim == 2:
        edges, simplices = get_2d_edges_and_simplices(
            x.shape, cyclical_axis=cyclical_axis
        )
    elif x.ndim == 3:
        edges, simplices = get_3d_edges_and_simplices(
            x.shape, cyclical_axis=cyclical_axis
        )
    else:
        raise ValueError("x must be 2D or 3D")
    psi = x.ravel()

    if weights is not None:
        weights = prepare_weights(weights, edges, merging_method)

    result = unwrap_arbitrary(
        psi, edges, None if use_edgelist else simplices, start_i=start_i,  weights=weights, **kwargs
    )
    if result is None:
        return None

    return result.reshape(x.shape)
```

**Comment by the-mysh** [comment]
Hi @yoyololicon , thanks for your review. I made changes according to your suggestions. I saw though that in case you use puma, the `weights` argument should not be present - so I opted for extracting it from kwargs and adding the prepared weights only if any were originally defined. 
Please let me know what you think!

File: kamui/__init__.py [hunk 3]
```diff
@@ -74,6 +76,10 @@ def unwrap_dimensional(
         raise ValueError("x must be 2D or 3D")
     psi = x.ravel()
 
+    weights = kwargs.pop('weights', None)
```
Comment (yoyololicon) [inline kamui/__init__.py:79]: I still think we should explicitly define the `weights` argument in the function signature and in the docstring, as I suggested in the previous review. 

`kwargs` here means the parameters for the core methods, which are edge-based. Explicitly defining and stating that `weights` should have the same shape as `x` helps people to know the it's node-based and is different. Also, in this way, `weights` will not be captured by `kwargs` and you don't have to call `pop` to get it out.

**Review by yoyololicon — COMMENTED** [review]
Just some minor issue.
Btw, can you attach some unwrapping examples to show how this vertex-based weighting works? Just curious. It's optional, though.

**Comment by yoyololicon** [comment]
@the-mysh please format the code using black.
See the failing workflow for details.

**Comment by the-mysh** [comment]
Hi @yoyololicon , sorry for the radio silence, got distracted by other work projects. I've implemented your change requests, please have a look when you get the chance.

File: kamui/__init__.py (hunk 3 above)
Comment (the-mysh) [reply kamui/__init__.py]: Here is an excerpt from the data I've been working on. The data comes from a cylindrical formation, so the second axis (azimuth in the picture) is cyclical. The weights come from an additional measurement and indicate how 'sure' we are about each pixel's value.

wrapped_and_weights

Here's some basic comparison of unwrapping without and with weights (converted to edge weights using the 'min' method). What we care about is that the inaccuracies are limited to smaller regions/are less extreme.

unwrapped

When using the edgelist method, the weights seem to have less influence on the final result.

unwrapped_edgelist

Comment (the-mysh) [reply kamui/__init__.py]: Overall, the weights give us some improvements for the most tricky parts of the data, but even without the weights `kamui` does a great job compared to our previous solutions, so thanks a lot for your work!

File: kamui/__init__.py [hunk 4]
```diff
@@ -124,6 +137,8 @@ def unwrap_arbitrary(
         The unwrapped phase of the same shape as psi.
     """
     if method == "gc":
+        if weights is not None:
+            raise ValueError("PUMA method does not take in weights")
```
Comment (yoyololicon) [inline kamui/__init__.py:141]: I prefer just let PUMA ignore `weights` as before.
```suggestion
```

File: kamui/__init__.py [hunk 5]
```diff
@@ -134,6 +149,7 @@ def unwrap_arbitrary(
                 np.round((psi[edges[:, 1]] - psi[edges[:, 0]]) / period).astype(
                     np.int64
                 ),
+                weights=weights,
```
Comment (yoyololicon) [inline kamui/__init__.py:152]:
```suggestion
```

File: kamui/__init__.py [hunk 6]
```diff
@@ -142,7 +158,7 @@ def unwrap_arbitrary(
             result = m * period + psi
         else:
             diff = wrap_difference(psi[edges[:, 1]] - psi[edges[:, 0]], period)
-            k = calculate_k(edges, simplices, diff / period, **kwargs)
+            k = calculate_k(edges, simplices, diff / period, weights=weights, **kwargs)
```
Comment (yoyololicon) [inline kamui/__init__.py:161]:
```suggestion
            k = calculate_k(edges, simplices, diff / period, **kwargs)
```

File: kamui/__init__.py [hunk 7]
```diff
@@ -45,6 +47,8 @@ def unwrap_dimensional(
     cyclical_axis : int or (int, int)
         The axis that is cyclical.
         Default to ().
+    weights :   Weights defining the 'goodness' of value at each vertex. Shape must match the shape of x.
```
Comment (yoyololicon) [inline kamui/__init__.py:50]:
```suggestion
    weights : Weights defining the 'goodness' of value at each vertex. Shape must match the shape of x.
```

File: kamui/__init__.py [hunk 8]
```diff
@@ -45,6 +47,8 @@ def unwrap_dimensional(
...
+    merging_method  : Way of combining two phase weights into a single edge weight.
```
Comment (yoyololicon) [inline kamui/__init__.py:51]:
```suggestion
    merging_method : Way of combining two phase weights into a single edge weight.
```

File: kamui/utils.py [hunk 9]
```diff
@@ -192,5 +196,67 @@ def get_3d_edges_and_simplices(
...
+) -> npt.NDArray[np.float_]:
+    """Prepare weights for `calculate_m` and `calculate_k` functions.
+
+    Assume the weights are the same shape as the phases to be unwrapped.
+
+    Scale the weights from 0 to 1. Pick the weights corresponding to the phase pairs connected by the edges.
+    Compute the mean/max/min (depending on the `merging_method`) of each of those pairs to give a weight for each edge.
+
+    Args:
+        weights         :   Array of weights of shapr corresponding to the original phases array shape.
```
Comment (yoyololicon) [inline kamui/utils.py:216]:
```suggestion
        weights         :   Array of weights of shape corresponding to the original phases array shape.
```

**Review by yoyololicon — APPROVED** [review]
There are a few lines I would like to change, but it looks good!
I'll add a few commits to address them, and then we can merge.

File: kamui/__init__.py (hunk 3 above)
Comment (yoyololicon) [reply kamui/__init__.py]: That's good to know!

**Comment by yoyololicon** [comment]
Thank @the-mysh, for the contribution! This will be included in the next release `v0.1.2`.

**Pull request merged**